File: behavioral/typescript/state_structured/vending_machine.ts
import { State } from './interfaces';

export class VendingMachine {
  private currentState: State;
  private coinCount: number = 0;
  private stockCount: number;
  
  constructor(initialStock: number) {
    this.stockCount = initialStock;
    if (initialStock > 0) {
      this.currentState = new NoCoinState(this);
    } else {
      this.currentState = new SoldOutState(this);
    }
  }
  
  setState(state: State): void {
    this.currentState = state;
  }
  
  insertCoin(): void {
    this.currentState.insertCoin();
  }
  
  ejectCoin(): void {
    this.currentState.ejectCoin();
  }
  
  pressButton(): void {
    this.currentState.pressButton();
  }
  
  dispense(): void {
    this.currentState.dispense();
  }
  
  addCoin(): void {
    this.coinCount++;
    console.log(`Coin inserted. Total coins: ${this.coinCount}`);
  }
  
  returnCoin(): void {
    if (this.coinCount > 0) {
      this.coinCount--;
      console.log(`Coin returned. Total coins: ${this.coinCount}`);
    }
  }
  
  releaseProduct(): void {
    if (this.stockCount > 0) {
      this.stockCount--;
      this.coinCount--;
      console.log(`Product dispensed. Stock: ${this.stockCount}, Coins: ${this.coinCount}`);
    }
  }
  
  getCoinCount(): number {
    return this.coinCount;
  }
  
  getStockCount(): number {
    return this.stockCount;
  }
}

export class NoCoinState implements State {
  private machine: VendingMachine;
  
  constructor(machine: VendingMachine) {
    this.machine = machine;
  }
  
  insertCoin(): void {
    console.log('Coin inserted...');
    this.machine.addCoin();
    this.machine.setState(new HasCoinState(this.machine));
  }
  
  ejectCoin(): void {
    console.log('No coin to eject');
  }
  
  pressButton(): void {
    console.log('Please insert a coin first');
  }
  
  dispense(): void {
    console.log('Please insert a coin first');
  }
}

export class HasCoinState implements State {
  private machine: VendingMachine;
  
  constructor(machine: VendingMachine) {
    this.machine = machine;
  }
  
  insertCoin(): void {
    console.log('Coin already inserted');
  }
  
  ejectCoin(): void {
    console.log('Coin ejected');
    this.machine.returnCoin();
    this.machine.setState(new NoCoinState(this.machine));
  }
  
  pressButton(): void {
    console.log('Button pressed...');
    if (this.machine.getStockCount() > 0) {
      this.machine.setState(new DispensingState(this.machine));
      this.machine.dispense();
    } else {
      console.log('Out of stock');
      this.machine.setState(new SoldOutState(this.machine));
    }
  }
  
  dispense(): void {
    console.log('Press button to dispense');
  }
}

export class DispensingState implements State {
  private machine: VendingMachine;
  
  constructor(machine: VendingMachine) {
    this.machine = machine;
  }
  
  insertCoin(): void {
    console.log('Please wait, dispensing...');
  }
  
  ejectCoin(): void {
    console.log('Cannot eject during dispensing');
  }
  
  pressButton(): void {
    console.log('Already dispensing...');
  }
  
  dispense(): void {
    this.machine.releaseProduct();
    if (this.machine.getStockCount() > 0) {
      this.machine.setState(new NoCoinState(this.machine));
    } else {
      this.machine.setState(new SoldOutState(this.machine));
    }
  }
}

export class SoldOutState implements State {
  private machine: VendingMachine;
  
  constructor(machine: VendingMachine) {
    this.machine = machine;
  }
  
  insertCoin(): void {
    console.log('Cannot insert coin, machine is sold out');
  }
  
  ejectCoin(): void {
    console.log('Cannot eject, no coin inserted');
  }
  
  pressButton(): void {
    console.log('Machine is sold out');
  }
  
  dispense(): void {
    console.log('Machine is sold out');
  }
}

File: behavioral/dart/state_structured/vending_machine.dart
import 'interfaces.dart';

class VendingMachine {
  late State _currentState;
  int _coinCount = 0;
  int _stockCount;
  
  VendingMachine(this._stockCount) {
    if (_stockCount > 0) {
      _currentState = NoCoinState(this);
    } else {
      _currentState = SoldOutState(this);
    }
  }
  
  void setState(State state) {
    _currentState = state;
  }
  
  void insertCoin() {
    _currentState.insertCoin();
  }
  
  void ejectCoin() {
    _currentState.ejectCoin();
  }
  
  void pressButton() {
    _currentState.pressButton();
  }
  
  void dispense() {
    _currentState.dispense();
  }
  
  void addCoin() {
    _coinCount++;
    print('Coin inserted. Total coins: $_coinCount');
  }
  
  void returnCoin() {
    if (_coinCount > 0) {
      _coinCount--;
      print('Coin returned. Total coins: $_coinCount');
    }
  }
  
  void releaseProduct() {
    if (_stockCount > 0) {
      _stockCount--;
      _coinCount--;
      print('Product dispensed. Stock: $_stockCount, Coins: $_coinCount');
    }
  }
  
  int get coinCount => _coinCount;
  int get stockCount => _stockCount;
}

class NoCoinState implements State {
  final VendingMachine _machine;
  
  NoCoinState(this._machine);
  
  @override
  void insertCoin() {
    print('Coin inserted...');
    _machine.addCoin();
    _machine.setState(HasCoinState(_machine));
  }
  
  @override
  void ejectCoin() {
    print('No coin to eject');
  }
  
  @override
  void pressButton() {
    print('Please insert a coin first');
  }
  
  @override
  void dispense() {
    print('Please insert a coin first');
  }
}

class HasCoinState implements State {
  final VendingMachine _machine;
  
  HasCoinState(this._machine);
  
  @override
  void insertCoin() {
    print('Coin already inserted');
  }
  
  @override
  void ejectCoin() {
    print('Coin ejected');
    _machine.returnCoin();
    _machine.setState(NoCoinState(_machine));
  }
  
  @override
  void pressButton() {
    print('Button pressed...');
    if (_machine.stockCount > 0) {
      _machine.setState(DispensingState(_machine));
      _machine.dispense();
    } else {
      print('Out of stock');
      _machine.setState(SoldOutState(_machine));
    }
  }
  
  @override
  void dispense() {
    print('Press button to dispense');
  }
}

class DispensingState implements State {
  final VendingMachine _machine;
  
  DispensingState(this._machine);
  
  @override
  void insertCoin() {
    print('Please wait, dispensing...');
  }
  
  @override
  void ejectCoin() {
    print('Cannot eject during dispensing');
  }
  
  @override
  void pressButton() {
    print('Already dispensing...');
  }
  
  @override
  void dispense() {
    _machine.releaseProduct();
    if (_machine.stockCount > 0) {
      _machine.setState(NoCoinState(_machine));
    } else {
      _machine.setState(SoldOutState(_machine));
    }
  }
}

class SoldOutState implements State {
  SoldOutState(VendingMachine machine);
  
  @override
  void insertCoin() {
    print('Cannot insert coin, machine is sold out');
  }
  
  @override
  void ejectCoin() {
    print('Cannot eject, no coin inserted');
  }
  
  @override
  void pressButton() {
    print('Machine is sold out');
  }
  
  @override
  void dispense() {
    print('Machine is sold out');
  }
}
